[WIP] MCH: digits merger

aferrero2707 opened the pull request:
This PR introduces the code required for the merging of SAMPA digits in continuous data taking mode, when the detector signals can be into multiple SAMPA acquisition windows that need to be combined in order to recover the full deposited charge.

The code is still work in progress, and depends on #3618 for the new digit time definition.

The PR introduces the following parts in `Detectors/MUON/MCH/Workflow`:
* a [DataDecoder](https://github.com/AliceMCH/AliceO2/blob/mch-digits-merger/Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx) class that runs the MCH raw data decoding routines on a CRU page, and merges the output digits when needed. The input/output interface of the data decoder is generic and can be used also in a non-DPL context
* a [spec class](https://github.com/AliceMCH/AliceO2/blob/mch-digits-merger/Detectors/MUON/MCH/Workflow/src/DataDecoderSpec.cxx) that provides the interfacing with the input/output DPL messages
* a [raw-to-digits](https://github.com/AliceMCH/AliceO2/blob/mch-digits-merger/Detectors/MUON/MCH/Workflow/src/raw-to-digits-workflow.cxx) workflow to insert the decoding step into a PDL pipeline

shahor02 commented:
Hi @aferrero2707, it seems your branch was not rebased to dev

aferrero2707 commented:
@shahor02 I will merge the current dev branch to fix the conflicts... I would not worry at this stage for the rather messy commit history, I will force-push a cleaner version once #3618 has been merged into dev.

In any case, the only files that are relevant to this PR are the `DataDecoder*` ones in `Detectors/MUON/MCH/Workflow`.

aferrero2707 commented:
@shahor02 @aphecetche is there a way to avoid triggering the CI jobs? They are not really useful at this stage of the PR, so it would be nice to spare some CPU time...

davidrohr commented:
@ktf : I was assuming the [WIP] in the title would prevent the CI from running? Isn't this the case anymore?

jgrosseo commented:
@davidrohr You have to create a draft PR (built in github)

aferrero2707 commented:
@jgrosseo the checks for this PR are still running, even if it is marked as draft and contains `[WIP]` in the title... maybe something else is needed to prevent the CI jobs from running?

aferrero2707 commented:
@aphecetche @pillot I have rebased the brach on the current `dev`. Now the number of changed files is much smaller and the commits should be much easier to read.

I still need to some basic unit tests for checking the digits merger code.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoderSpec.cxx:
```diff
@@ -56,165 +57,30 @@ using namespace o2::framework;
   //_________________________________________________________________________________________________
-  void
-    decodeTF(framework::ProcessingContext& pc, std::vector<o2::mch::Digit>& digits)
+  void decodeTF(framework::ProcessingContext& pc)
```
I believe in this method you can safely replace RAWDataHeaderV4 by RAWDataHeader and thus get the code RDH version agnostic

aferrero2707 replied:
Done.

aphecetche commented on Detectors/MUON/MCH/Workflow/include/MCHWorkflow/DataDecoder.h:
```diff
@@ -0,0 +1,132 @@
+// or submit itself to any jurisdiction.
+
+///
+/// \file    DatDecoder.cxx
```
typo DatDecoder -> DataDecoder

aferrero2707 replied:
Fixed.

aphecetche commented on Detectors/MUON/MCH/Workflow/include/MCHWorkflow/DataDecoder.h:
```diff
@@ -0,0 +1,132 @@
+/// \brief Implementation of a data processor to run the raw decoding
+///
+
+#include <random>
```
Does the DataDecoder.h needs to be public (i.e. in include/MCHWorkflow) ? 
If it does (and even if it does not, by the way), you should limit as much as possible the number of includes it uses. Do you need as many ? Move as much as possible in the .cxx.

aferrero2707 replied:
I guess It needs if we want to use the decoder independently of the DPL, for example when processing the raw data directly in the QC.

I have removed all the unnecessary headers.

aphecetche commented on Detectors/MUON/MCH/Workflow/include/MCHWorkflow/DataDecoder.h:
```diff
@@ -0,0 +1,132 @@
+#include "MCHRawCommon/DataFormats.h"
+#include "MCHRawDecoder/PageDecoder.h"
+#include "MCHRawElecMap/Mapper.h"
+//#include "MCHRawCommon/RDHManip.h"
```
remove commented code

aferrero2707 replied:
Done.

aphecetche commented on Detectors/MUON/MCH/Workflow/include/MCHWorkflow/DataDecoder.h:
```diff
@@ -0,0 +1,132 @@
+//#include "MCHRawCommon/RDHManip.h"
+#include "MCHMappingInterface/Segmentation.h"
+#include "MCHWorkflow/DataDecoderSpec.h"
+//#include <array>
```
remove commented code

aferrero2707 replied:
Done.

aphecetche commented on Detectors/MUON/MCH/Workflow/include/MCHWorkflow/DataDecoder.h:
```diff
@@ -0,0 +1,132 @@
+
+#define MCH_MERGER_FEEID_MAX 63
+
+namespace o2::header
```
set other comments about RDH versioning. Mention of a specific version should no longer be needed ?

aferrero2707 replied:
I have removed this part, it was not really neeed.

aphecetche commented on Detectors/MUON/MCH/Workflow/include/MCHWorkflow/DataDecoder.h:
```diff
@@ -0,0 +1,132 @@
+namespace raw
+{
+
+using namespace o2;
```
no using namespace in include files, please. (yes, we have a couple in MUON/MCH/Raw that I should remove as well ;-) )

aferrero2707 replied:
All removed!

aphecetche commented on Detectors/MUON/MCH/Workflow/include/MCHWorkflow/DataDecoder.h:
```diff
@@ -0,0 +1,132 @@
+//_________________________________________________________________
+class DataDecoder
+{
+ private:
```
that's of course a matter of taste, but I prefer to put the public interface first, as at first glance I don't care about the private stuff.

aphecetche commented on Detectors/MUON/MCH/Workflow/include/MCHWorkflow/DataDecoder.h:
```diff
@@ -0,0 +1,132 @@
+  void initElec2DetMapper(std::string filename);
+  void initFee2SolarMapper(std::string filename);
+
+ public:
```
that's a recurring pattern I see in some workflows : why don't you provide a proper constructor of the DataDecoder class instead of having a default (and useless) one plus a bunch of setXXX methods ? Makes it more difficult to understand what's needed to get a working DataDecoder built IMO

aferrero2707 replied:
Done. All setXXX methods are replaced by a constructor.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoderSpec.cxx:
```diff
@@ -56,165 +57,30 @@ using namespace o2::framework;
+    auto mDs2manu = ic.options().get<bool>("ds2manu");
     mPrint = ic.options().get<bool>("print");
 
+    mDecoder.setDs2manu(mDs2manu);
```
see my comment above about regular constructor vs bunch of set methods

aferrero2707 replied:
See above.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoderSpec.cxx:
```diff
@@ -264,56 +130,43 @@ class DataDecoderTask
-
-    decodeTF(pc, digits);
+    mDecoder.reset();
+    //std::cout << "DatDecoderTask::run()" << std::endl;
```
please remove commented out code

aferrero2707 replied:
Done.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoderSpec.cxx:
```diff
@@ -233,12 +99,12 @@ class DataDecoderTask
 
   //_________________________________________________________________________________________________
-  void decodeReadout(const o2::framework::DataRef& input, std::vector<o2::mch::Digit>& digits)
+  void decodeReadout(const o2::framework::DataRef& input)
```
now I am confused : what is the "readout" input ? Please add some doc

aferrero2707 replied:
I have added some comments as well as an example in the README.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/cru-page-to-digits-workflow.cxx:
```diff
@@ -0,0 +1,39 @@
+///
+/// \brief This is an executable that runs the decoding via DPL.
+///
+/// This is an executable that takes a raw buffer from the Data Processing Layer, runs the decoding and sends the digits via the Data Processing Layer.
```
I guess this is a cut and paste from the raw-to-digits workflow.
BTW, please add in the README some explanation about the difference about the two

aferrero2707 replied:
See above.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx:
```diff
@@ -0,0 +1,602 @@
+using namespace o2::mch::mapping;
+using RDHv4 = o2::header::RAWDataHeaderV4;
+
+static std::array<int, 64> refManu2ds_st345_v5 = {
```
please add some comments explaining the difference between those two arrays

aferrero2707 replied:
Done.

aphecetche commented:
@aferrero2707 I've started to review the code, but still have to go through the core of it (MergerXXX classes) ... Will do that this afternoon hopefully

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx:
```diff
@@ -0,0 +1,602 @@
+  return refDs2manu_st345[i];
+}
+
+class MergerDigit
```
I'm confused by the name here : should'nt it be Merge**d**Digit instead ?

aferrero2707 replied:
No, this structure does not represent an already merged digit, instead it represents a single un-merged digit with some additional informations needed by the merging algorithm.

aphecetche replied:
That's a tad confusing, but ok.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx:
```diff
@@ -0,0 +1,602 @@
+  void mergeDigits();
+};
+
+//_________________________________________________________________
```
I guess the MergerXXX (would have called them XXXMerger instead, btw) are important enough to live in their own files (.h / .cxx). Will ease testing as well...

aferrero2707 replied:
Indeed. I have moved the code in separate source files and changed the name of the classes as you propose.

aphecetche replied:
Thanks, but I'll have some comments on the new file (DigitsMerger.cxx) ;-)

ktf commented:
@davidrohr (sorry I missed this). AFAICT, the PR checks did not run, only the travis one, which we cannot "hold".

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx:
```diff
@@ -0,0 +1,602 @@
+
+class MergerDigit
+{
+ public:
```
if it's meant to stay as simple, then a class with only public stuff = a struct (and can then remove the ctor mentions)

aferrero2707 replied:
Done.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx:
```diff
@@ -0,0 +1,602 @@
+    if (d1.merged || d2.merged)
+      return false;
+
+    // skip all digits not matching the detId/padId
```
I would externalise this check to e.g. a free function (bool areSamePad(const MergedDigit& d1, const MergedDigit& d2) for instance) or a comparison operator of the MergedDigit

aferrero2707 replied:
Done.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx:
```diff
@@ -0,0 +1,602 @@
+    if (d1.digit.getPadID() != d2.digit.getPadID())
+      return false;
+
+    // compute time difference
```
here also I would externalise this time difference computation's logic (and use the BXTime digit function within it) (and maybe it be also be reused somehow in the other PR regarding time clustering btw ?)

aferrero2707 replied:
Here I would not touch this part of the code yet. I have in mind to use the bunch crossing value at the beginning of the Time Frame (taken from dedicated Heartbeat trigger events at the beginning of the TF) in order to remove the need for the rollover correction. However I still have to think carefully about how to implement it. The same logic could then be used for the comparisons in the time clustering code...

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx:
```diff
@@ -0,0 +1,602 @@
+    }
+
+    // skip if the time difference is not equal to 1 ADC clock cycle
+    if (timeDiff > 8)
```
get this 8 into a constant or a parameter of some sort

aferrero2707 replied:
The check is now updated to `if (timeDiff != 1)`. The `8` was due to a bug in the configuration of the SAMPA chips in the data used for the tests. Shall we put `1` into a constant as well?

aphecetche replied:
Yes, please, e.g. ONE_ADC_CLOCK or something alike. And yes, I know currently there's a comment above that line stating this, but comments are notorious to get out of sync with the code, so it never hurts to be crystal clear in the code itself, IMO

aferrero2707 replied:
It is there now: https://github.com/AliceMCH/AliceO2/blob/mch-digits-merger/Detectors/MUON/MCH/Raw/Decoder/src/DigitsMerger.cxx#L87

aphecetche replied:
@aferrero2707 sorry I was indeed looking at a previous version of the PR. I've just updated it so hopefully the rest of the comments will be on up-to-date code ;-)

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx:
```diff
@@ -0,0 +1,602 @@
+    if (timeDiff > 8)
+      return false;
+
+    // merge digits
```
provide a += operator (or some e.g. mergeWith member function) to the MergedDigit ? to make it clearer which of the two will be modified ? 
```
d2 += d1;
d1.merged = true;
```

aferrero2707 replied:
Done

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx:
```diff
@@ -0,0 +1,602 @@
+    uint32_t bxStop = stopTime.bunchCrossing;
+    // correct for value rollover
+    if (bxStart < bxStop)
+      bxStart += 0x100000;
```
get this 0x100000 into a constant or a parameter of some sort

aferrero2707 replied:
Done.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx:
```diff
@@ -0,0 +1,602 @@
+
+void FeeIdMerger::mergeDigits()
+{
+  auto checkDigits = [](MergerDigit& d1, MergerDigit& d2) -> bool {
```
that may be just me but when I see a function called checkXXX I do not expect it to modify its parameters. But this one does (ok, the parameters are passed by reference which is a hint for the reader...)

aferrero2707 replied:
You are right. I have renamed it to `updateDigits`.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx:
```diff
@@ -0,0 +1,602 @@
+
+    int padId = -1;
+    const Segmentation& segment = segmentation(deId);
+    if ((&segment) == nullptr) {
```
I believe the call to segmentation(invalid_deid) will simply crash, so no need for a check. Or said otherwise, the check of deId validity should be made beforehand

aferrero2707 replied:
Is there an existing function for checking the validity of a DetID? Or could this be added to the MCH mapping library? This is a recurring problem in different places, like the QC...

aphecetche replied:
Not quite yet, but there's a deIdsForAllMCH array defined in MCHRaw/Mapper.h that could easily be used to define such a function. I'll have a look.

aferrero2707 replied:
Ok, then I'll use the deIdsForAllMCH array for the beginning, and we can introduce the function when available.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx:
```diff
@@ -0,0 +1,602 @@
+      channel = ds2manu(int(channel));
+    }
+
+    double digitadc(0);
```
could probably be replaced by a float chargeSum(const SampaCluster& sc) function in SampaCluster.h/.cxx ? btw what happens if this code is ran on data taken in chargesum mode ?

aferrero2707 replied:
Done. The code now also handles the chargeSum case properly.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx:
```diff
@@ -0,0 +1,602 @@
+      return;
+    }
+
+    Digit::Time time;
```
here again are opportunities to a couple free functions to make the code clearer ? 
e.g. 
```
Digit::Time startTime(const SampaCluster& sc, uint32_t orbit) { return Digit::Time{sc.sampaTime,sc.bunchCrossing,orbit); }
Digit::Time stopTime(const SampaCluster& sc, uint32_t orbit) { return Digit::Time{sc.sampaTime+sc.nofSamples()-1,sc.bunchCrossing,orbit); }
```

aferrero2707 replied:
Again, I would propose to wait for a possible improved definition of the digit time based on the bunch crossing value at the beginning of each Time Frame.

aphecetche replied:
well, seeing that actually the only thing needed for the client code (looking here at DigitsMerger.cxx L121) is the timeDiff, I would write a timeDiff(const Digit& d1, const Digit& d2) and let the implementation of that timeDiff change if need be later on (including the removal of the printouts at some point) ?  what do you think ?

aferrero2707 replied:
Good, point, I have introduced the timeDiff function as you propose.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx:
```diff
@@ -0,0 +1,602 @@
+
+  const auto patchPage = [&](gsl::span<const std::byte> rdhBuffer) {
+    auto rdhAnyPtr = reinterpret_cast<o2::header::RDHAny*>(const_cast<std::byte*>(&rdhBuffer[0]));
+    if (mRdhHandler) {
```
is this RDHHandler actually used somewhere ?

aferrero2707 replied:
I plan to use it for example in the QC, when directly processing the raw data using the O2 decoder. That's actually the reason why I have introduced it in the first place, such that code using the decoder can still access the sequence of RDHs in the input buffer without having to walk through the data.

What do you think?

aphecetche replied:
I always find a bit tricky/risky to introduce code that is not-yet-used (as it's difficult to know beforehand if e.g. it's the right interface, and if/when it will actually be used), but OK, I see your point. Fine with me.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx:
```diff
@@ -0,0 +1,602 @@
+    ++ndigits;
+  };
+
+  const auto patchPage = [&](gsl::span<const std::byte> rdhBuffer) {
```
the patchPage is now doing more that what it says on the tin : it patches the RDH (expected), handle the RDH (assuming there's a rdh handler defined, might be expected, not so sure) and "configure" the merger (unexpected).

aferrero2707 replied:
I agree that the pieces of code that are doing more than patching the RDH contents should be moved outside. I have already created a function for the merger initialization and update, while for the orbit info I propose to wait until #4285 is merged and then port the corresponding code.

aphecetche replied:
hum, ok, but then it means #4285 must go first and then this one be rebased on it ?

aferrero2707 replied:
Yes... in any case, I consider #4285 more urgent that this one.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx:
```diff
@@ -0,0 +1,602 @@
+    if (mRdhHandler) {
+      mRdhHandler(rdhAnyPtr);
+    }
+    auto rdhPtr = reinterpret_cast<o2::header::RAWDataHeaderV4*>(const_cast<std::byte*>(&rdhBuffer[0]));
```
should be replaced by 
```
const void* rdhP = reinterpret_cast<const void*>(rdhBuffer.data());
```
to get a RDH-version agnostic code

shahor02 replied:
Actually, what is the point of having ``o2::header::RDHAny* rdhAnyPtr`` a few lines above and then the same pointer as ``RAWDataHeaderV4`` (or even ``void`` as @aphecetche suggest?).
It seems it is never used, as well as its dereferenced version on L464.
In general, I would suggest to use the reference on RDHAny, i.e.
````
using RDH = o2::header::RDHAny;
....
auto& rdhAny = *reinterpret_cast<RDH*>(const_cast<std::byte*>(&rdhBuffer[0]));
````
This will provide the same RDHUtils functionality as using the pointer (all methods are interfaced to both pointers and refs),  but has an advantage that once we will settle on the final header, we can alias RDH to this header. Then, the tiny overhead we can have in RDHUtils due to the version check will be fully avoided.

aferrero2707 replied:
I have adopted the suggestion from @shahor02, hopefully in all relevant places.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx:
```diff
@@ -0,0 +1,602 @@
+      mRdhHandler(rdhAnyPtr);
+    }
+    auto rdhPtr = reinterpret_cast<o2::header::RAWDataHeaderV4*>(const_cast<std::byte*>(&rdhBuffer[0]));
+    auto& rdh = *rdhPtr;
```
not used. remove.

aferrero2707 replied:
Done.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx:
```diff
@@ -0,0 +1,602 @@
+      std::cout << fmt::format("\tC {}  PAD_XY {:+2.2f} , {:+2.2f}", (int)bend, X, Y);
+      std::cout << std::endl;
+    }
+    for (auto d : mOutputDigits) {
```
certainly the printing of the bending/non-bending digits can be done without that much duplicated code ? (one more free function, e.g. dumpDigit(gsl::span<Digit> digits, bool bending) to be done ?)

aferrero2707 replied:
The double loop is there to allow printing all the bending digits first, then all the non-bending ones, to simplify the debugging.  I have improved the code with an helper function.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx:
```diff
@@ -0,0 +1,602 @@
+      std::cout << fmt::format("\tC {}  PAD_XY {:+2.2f} , {:+2.2f}", (int)bend, X, Y);
+      std::cout << std::endl;
+    }
+    for (auto d : mOutputDigits) {
```
in principle this final loop should be empty, right ? No digit should have an invalid padId ?

aferrero2707 replied:
Definitely. I must have introduced it for debugging purposes. I removed it.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx:
```diff
@@ -0,0 +1,602 @@
+
+//=======================
+// Data decoder
+void DataDecoder::decodeBuffer(gsl::span<const std::byte> page)
```
this function is quite long and difficult to follow as it is now

aferrero2707 replied:
You are right, I will try to improve it. However I am not sure we can make the channel handler much shorter...

aphecetche replied:
Agreed. For now, and for the channelHandler at least ;-) (as I don't have immediate proposal on how to improve it) 

I still believe though that 4 lambdas in this method is a bit much and makes the code quite hard to read.  Probably 2/4 could be externalised : e.g. the dumpDigits could be a simple free function take a vector of digits as argument, and the updateMerger, well, I'd have to think about it ;-) )

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx:
```diff
@@ -0,0 +1,602 @@
+//_________________________________________________________________
+class FeeIdMerger
+{
+  MergerBuffer buffers[2];
```
Instead of having to juggle with two buffers indices, why not define two actual buffers named currentBuffer and previousBuffer and use a swap method to revert them when needed (if I understand correctly is what's done below in L230 ?). Then you can simplify and remove the BufId (and their get methods) altogether

aferrero2707 replied:
Good suggestion. I will change the code to use `std::swap`.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx:
```diff
@@ -0,0 +1,602 @@
+  virtual void mergeDigits(int feeId) = 0;
+};
+
+class MergerSimple : public MergerBase
```
if I understand correctly the simple merger is actually not merging at all, right ?

aferrero2707 replied:
Right. We currently only merge data coming from the User Logic, and we should actually only do that when running in continuous mode. In all other cases one can spare some CPU time by simply bypassing the merging algorithm.

aphecetche replied:
ok, then it's more than a SimpleMerger, it's a NoOpMerger ;-)

aferrero2707 replied:
I renamed the class NoOpMerger for clarity...

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx:
```diff
@@ -0,0 +1,602 @@
+  int nSent = 0;
+  for (auto& d : buffers[previousBufId].digits) {
+    if (!d.merged && (d.digit.getPadID() >= 0)) {
+      uint16_t size = (d.stopTime.getBXTime() - d.digit.getTime().getBXTime()) / 4 + 1;
```
not used . remove

aferrero2707 replied:
Done.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx:
```diff
@@ -0,0 +1,602 @@
+
+void Merger::setOrbit(int feeId, uint32_t orbit, bool stop)
+{
+  if (feeId < 0 || feeId > MCH_MERGER_FEEID_MAX) {
```
is this check actually needed in all the merger's functions ? shouldn't the feeid be checked upfront somehow ?

aferrero2707 replied:
Not really. I have added it in order to have some extra checks during the debugging. I propose to remove it once we are ready to drop the WIP label.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx:
```diff
@@ -0,0 +1,602 @@
+    std::cout << "Merging digits in " << previousBufId << " (orbit=" << previousBuffer.orbit << ")\n";
+  }
+
+  for (size_t i = 0; i < previousBuffer.digits.size(); i++) {
```
I can't say I understand why this loop is there. The second one below which merges current with previous I think I get, but not this one... Care to explain ?

aferrero2707 replied:
Sure... let me try to briefly outline the whole data flow. If the explanation is clear enough we might also put it as documentation in the source file.

First of all, during the processing of orbit N the corresponding digits are added to the current merge buffer (there is one merge buffer for each FEEID).

When the end of orbit N is reached (either because we receive the stop RDH or because the orbit number increases), the merging is performed. First, the not-yet-merged digits in the previous buffer are merged together, then the merging between the previous and current buffers is performed. The "seeds" of the merging algorithm are digits for which the first ADC sample is exactly at the beginning of the SAMPA time window, i.e. the SAMPA "cluster time" is equal to zero. Since we are only merging digits belonging to the same SAMPA channel, and that the data coming from each SAMPA channel is already in time-ascending order, a single loop on the seeds followed by an inner loop on the non-seed digits is enough to catch all combinations.

When the merging of the two buffers is completed, the merged digits of the previous buffer are sent out. There is therefore an offset of one orbit between the raw decoding and the digits output. This in particular means that the digits belonging to the last orbit of a given input buffer (typically a TimeFrame) will not be sent to the output, because they would need to be merged with the those from the first orbit of the next TF.

Does it clarify your doubts?

aphecetche commented:
@aferrero2707 I think I'm done with the review for now. Sorry for the numerous comments, hope it's still readable...

ktf commented:
This is currently not merging and still WIP. Is it still being worked on?

aphecetche commented:
@aferrero2707 what is the status of this one ?

aferrero2707 commented:
@aphecetche I have just finished rebasing the branch to the current official `dev` branch, and I am now checking that everything is OK after the changes.

I will then go through your comments and do the necessary changes. The work on this actually stopped because we resumed the activities at P2... now I am trying to catch up.

The main missing part for the moment is the implementation of appropriate tests for the merging code. I propose to keep the WIP label until this is added.

aferrero2707 committed:
[MCH] add digits merger workflow

The digits merger combines SAMPA digits that span multiple time windows.

aferrero2707 commented:
@aphecetche I have implemented all the changes discussed in the review, and force-pushed a new version with all commits combined, for better readability.

aphecetche commented:
@aferrero2707 has this PR been rebased on dev ?

aferrero2707 commented:
@aphecetche in principle yes, 3 days ago... unless I did some mistake!

aferrero2707 committed:
[MCH] added overloaded operator== for Digt::Time structure

aferrero2707 committed:
[MCH] improvements to DigitsMerger code

- removed redundant stopTime member to MergerDigit structure
- added explicit previous and current merge buffers
- fixed bug in computation of time difference condition for digits merging

aferrero2707 committed:
[MCH] added initial unit test for digits merger

aferrero2707 committed:
[MCH] improved unit tests fro digits merger

aferrero2707 commented:
@aphecetche @ktf any idea why this has been automatically closed despite the recent activities?

ktf commented:
Not really, but the WIP might have something to do with it. I will try to check.

aferrero2707 commented:
@ktf thanks!

aferrero2707 committed:
[MCH] added validity check for detection element IDs

aferrero2707 committed:
[MCH] removed MergerDigits struct

aferrero2707 committed:
[MCH] renamed SimpleMerger as NoOpMerger

aferrero2707 committed:
[MCH] further improvements to merger code readibility
